### DISCUSSION TITLE
Liunx DRM Shared CancellationToken

### DISCUSSION THREAD

**Discussion by workgroupengineering** [AvaloniaUI/Avalonia#15033] [Q&A] [answered]
I need share CancellationToken beetwenn Avalonia DRM app and background service. From what I've seen it's not possible at the moment. 
I would like to create a PR but I have doubts about which implementation to choose. Some possibilities are:

- Adding CancellationToken parameter to StartLinuxDirect 
- Adding CancellationToken to Linux FramebufferPlatformOptions
- Using AppBuilerd.With to registry CancellationToken

In your opinion, what is the best solution?

**Comment by kekekeks** [comment]
We probably want to expose a way to create a TopLevel from IInputBackend/IOutputBackend, which would allow skipping lifetime API usage, so one could just pass their own token to MainLoop. Should also enable multiseat support _of sorts_, if device has multiple displays connected.

**Reply by kekekeks** [reply, marked as answer]
As a workaround, you can do something like this:

```cs
            builder.AfterSetup(b =>
            {
                var controlled = (IControlledApplicationLifetime)b.Instance!.ApplicationLifetime;
                cancel.Register(() => Dispatcher.UIThread.Post(() => controlled.Shutdown(), DispatcherPriority.Send)));
            };
```
